### PR TITLE
Add unit tests for com.amazon.opendistroforelasticsearch.security.com…

### PR DESCRIPTION
**Pull request by Braavos6**
I've analysed your codebase and noticed that com.amazon.opendistroforelasticsearch.security.compliance.ComplianceConfig is not fully tested.
I've written some tests for the methods in this class with the help of [Diffblue Cover](https://www.diffblue.com/opensource).

Hopefully, these tests will help you detect any regressions caused by future code changes. If you would find it useful to have additional tests written for this repository, I would be more than happy to look at other classes that you consider important.

### CODE REVIEW

File: src/test/java/com/amazon/opendistroforelasticsearch/security/compliance/ComplianceConfigTest.java [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+        objectUnderTest.setExternalConfigLogged(false);
+        Reflector.setField(objectUnderTest, "logExternalConfig", false);
+
+        //for extra coverage calling getters
+        Assert.assertNull(objectUnderTest.getSettings());
+        Assert.assertNull(objectUnderTest.getEnvironment());
+
+        Assert.assertFalse(objectUnderTest.isLogExternalConfig());
+        Assert.assertFalse(objectUnderTest.isExternalConfigLogged());
+        Assert.assertFalse(objectUnderTest.isEnabled());
```
Comment (freddytuxworth) [inline src/test/java/com/amazon/opendistroforelasticsearch/security/compliance/ComplianceConfigTest.java]: What does this test?

File: src/test/java/com/amazon/opendistroforelasticsearch/security/compliance/ComplianceConfigTest.java [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+        Assert.assertFalse(objectUnderTest.isIndexImmutable("foo"));
+    }
+
+    @Test
+    public void testGetSalt16() throws InvocationTargetException {
+        ComplianceConfig objectUnderTest = (ComplianceConfig) Reflector.getInstance(
+                "com.amazon.opendistroforelasticsearch.security.compliance.ComplianceConfig");
+        Reflector.setField(objectUnderTest, "salt16", new byte[]{});
+
+        Assert.assertArrayEquals(new byte[]{}, objectUnderTest.getSalt16());
```
Comment (freddytuxworth) [inline src/test/java/com/amazon/opendistroforelasticsearch/security/compliance/ComplianceConfigTest.java]: This test just assigns a field reflectively and tests the getter

File: src/test/java/com/amazon/opendistroforelasticsearch/security/compliance/ComplianceConfigTest.java (hunk 1 above)
Comment (Braavos6) [reply src/test/java/com/amazon/opendistroforelasticsearch/security/compliance/ComplianceConfigTest.java]: //for extra coverage calling getters
I have removed them

File: src/test/java/com/amazon/opendistroforelasticsearch/security/compliance/ComplianceConfigTest.java (hunk 2 above)
Comment (Braavos6) [reply src/test/java/com/amazon/opendistroforelasticsearch/security/compliance/ComplianceConfigTest.java]: Removed test case

File: pom.xml [hunk 3]
```diff
@@ -230,6 +230,29 @@
...
+    <dependency>
+      <groupId>org.mockito</groupId>
+      <artifactId>mockito-all</artifactId>
+      <version>1.10.19</version>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>com.diffblue</groupId>
+      <artifactId>deeptestutils</artifactId>
+      <version>1.8.1</version>
```
Comment (JohnLBergqvist) [inline pom.xml]: use version 1.9.0 of `deeptestutils` and 1.6.6 of both `powermock` dependencies

**Review by JohnLBergqvist — CHANGES_REQUESTED** [review]
Can you justify the use of reflection & Mocking?

**Comment by Braavos6** [comment]
> Can you justify the use of reflection & Mocking?

Yes. It was easier than creating objects (Environment, IndexResolverReplacer, AuditLog) for the constructor of ComplianceConfig, which would have taken more lines of code.

File: pom.xml (hunk 3 above)
Comment (Braavos6) [reply pom.xml]: Versions changed

**Comment by JohnLBergqvist** [comment]
> > Can you justify the use of reflection & Mocking?
> 
> Yes. It was easier than creating objects (Environment, IndexResolverReplacer, AuditLog) for the constructor of ComplianceConfig, which would have taken more lines of code.

I'm still not sure this is the best approach, but i'll ask @paulkennethkent what he thinks

**Comment by paulkennethkent** [comment]
> > > Can you justify the use of reflection & Mocking?
> > 
> > 
> > Yes. It was easier than creating objects (Environment, IndexResolverReplacer, AuditLog) for the constructor of ComplianceConfig, which would have taken more lines of code.
> 
> I'm still not sure this is the best approach, but i'll ask @paulkennethkent what he thinks

I don't think this is the best approach as you are missing on some of the logic of the constructor. 
I think the bigger issue is the amount of reflection these tests have.

**Pull request closed without merging**